Fixed issue #10759 :  Added --striper option to rados command line

sponce opened the pull request:
Although libradosstriper has been introduced in giant, there has been no way to use it from the command line via the rados command. This is now available via --striper flag.

dachary commented:
could you add tests that uses these flags and verifies the intended behavior ? Verifying is probably tricky because it's supposed to be transparent. osd/osd-scrub-repair.sh is a good example to follow for such a test. It runs a tiny cluster on which you can run rados.

sponce commented:
I've now added a test, but I need you to tell me why it gives only 'operation not supported' when I try to run it standalone, while it succeeds when I run its core against a regular cluster.
I've also cleaned up the log message.

dachary commented on src/test/libradosstriper/rados-striper.sh:
```diff
@@ -0,0 +1,89 @@
+    local dir=$1
+    shift
+
+    export CEPH_MON="127.0.0.1:7107"
```
s/7107/7112/ because it's not used by any other test and will allow them to run in //

sponce replied:
done

dachary commented:
@sponce can you show me the output of the test please ?

sponce commented:
The output is long, but here is the relevant part when it fails is here :

run: 35: dd if=/dev/urandom of=testdir/rados-striper/toyfile bs=1234 count=1
1+0 records in
1+0 records out
1234 bytes (1.2 kB) copied, 0.000183255 s, 6.7 MB/s
run: 38: rados --pool rbd --striper put toyfile testdir/rados-striper/toyfile
error putting rbd/toyfile: (95) Operation not supported
run: 38: return 1

Testing by hand and attaching gdb, what I see is that the locking is returning the error. Here are the lines in the libradosstriper code failing : 
```c++
  librados::ObjectWriteOperation op;
  op.assert_exists();
  m_lockCookie = RadosStriperImpl::getUUID();
  utime_t dur = utime_t();
  rados::cls::lock::lock(&op, RADOS_LOCK_NAME, LOCK_EXCLUSIVE, m_lockCookie, "", "", dur, 0);
  int rc = m_ioCtx->operate(oid, &op);
```
rc is -95 !
The same line is working fine when it's not  the toy cluster of the test.

dachary commented:
the -95 is strange. Do you see something relevant in the logs ? The mini cluster logs are removed at the end of the test. But if you exit 0 immediately after the error instead of returning, you will find them in testdir.

sponce commented:
Thanks for that tip. That's all I needed. I've been able to run now, but I needed to do some strange symlink. The error I found was :

2015-02-06 14:49:25.051883 7f10ca1e9700  0 _load_class could not stat class /usr/local/lib/rados-classes/libcls_lock.so: (2) No such file or directory

The point is that I was running from a fresh compilation located in some weird path, and I did set LD_LIBRARY_PATH accordingly. It worked for all the rest, but somehow it was not respected in this case, leading to not find libcls_lock.so.
Adding a symlink in /usr/local/lib for my weird path allowed me to run.
Any clue what is wrong ?

At lest for this pool request, I'm happy : the test works

dachary commented:
I think there is a config_opts that should be set to point to the right directory when writing from sources.

dachary commented:
<pre>
--- a/src/test/ceph-helpers.sh
+++ b/src/test/ceph-helpers.sh
@@ -435,4 +435,5 @@ function activate_osd() {
     ceph_args+=" --chdir="
     ceph_args+=" --osd-pool-default-erasure-code-directory=.libs"
+    ceph_args+=" --osd-class-dir=.libs"
     ceph_args+=" --run-dir=$dir"
     ceph_args+=" --debug-osd=20"
</pre>
should help. If it does, would you mind adding this fix as a separate commit ? (not a separate pull request, just a separate commit will be enough ;-)

sponce commented:
I knew it would be an easy one for you :-)
It works and I've added the extra commit.

dachary commented:
:-) Now that it works, could you add src/test/libradosstriper/rados-striper.sh to the list of tests that are running with make check in Makefile.am ? If you look for osd-scrub-repair.sh you'll find an appropriate place.

sponce commented:
Done. I hope it's the right place as osd-scrub-repair.sh is nowhere. It's actually called from others which I found where I've put mine.

dachary commented:
This is the right place. And I'll add the osd-scrub-repair.sh script to the tests, thanks for finding it was never run. Do you mind waiting a few days for this pull request to be merged ? Just to let the hammer branch settle before adding new features to master. Just in case something's missing, not having new features will make things easier.

sponce commented:
Great. And no problem to wait a bit.

dachary commented:
rebased & reworked at #4018
